Add "Organization" tab with a diagram

K-Phoen opened the pull request:
## Hey, I just made a Pull Request!

I'm opening this PR early (changeset and tests not written yet) to get some early feedback.

I was playing around with having the org chart generated automatically based on Groups in the catalog, and I have a few questions:
* is there any interest in having this in Backstage?
* if so, where should it live? For my use case, I'd like to have it accessible via the "explore" plugin and since the tabs displayed on the explore page are static, this means I had to tweak the plugin itself
* are there any plans to make the tabs in the explore plugin customizable? Like what #2791 is doing for the catalog plugin.

This is how it currently looks:

![org_diagram](https://user-images.githubusercontent.com/66958/116010059-2205cd00-a61d-11eb-9aa6-005d3042c271.png)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Fox32 commented:
This could also live as a widget inside the `org` plugin. You could conditionally display it on group entity pages of type `root` (the root group entity on an organization). Maybe you event want to have a diagram that shows the context around a group (parent and child groups) and allows to expand to the whole organization.

adamdmharvey reviewed (commented):
I think the idea is awesome, so nice contrib @K-Phoen !! 👏🏻 

A thought from my use of Backstage, and of course each company is different, but a few places I've worked where Backstage would have been used would more likely have intended to expose Backstage groups from GitHub teams.  Given this, GitHub teams at large companies I've found are not a reflection of our organizational structure by design. That is, we may have reporting structures, managers, SAP/Workday reporting structure teams, but dynamic squads made up of members from different teams are pretty common, as well as code or libraries owned by a group of like minded individuals from different teams who may not even work together on a daily basis.  Given this, calling the diagram the "organization chart" wouldn't be accurate for us.  But and I think @Fox32 starts to elude to it above, using the visualization to see the team relationships is hugely valuable. Especially because I think it helps draw out best practice of true IAM-style team membership which a lot of new (or rushed) teams skip over. (oh i'll just add one name here, and another name here, anddddd wham - 2 years later we still only have assigned named without teams so onboarding new team members is really hard, or changing team make up if squads change)

Anyway, my feedback is simply that presenting such a diagram of Backstage teams is hugely valuable and I love the visualization, but it might not truly be a reflection of a company's organization structure per se, but a reflection of their team/group memberships.

adamdmharvey commented:
Final point: this could be amazing at drawing out [Conway's Law](https://en.wikipedia.org/wiki/Conway%27s_law) type conclusions, where engineering leadership may BELIEVE something is structured in one way, but SEEING it in this type of chart may help prove where the product boundaries really are and thus help them make different strategic decisions about how to re-org their squads/teams!

regicsolutions commented:
This looks awesome. Another use case that would be cool to see for larger orgs is a view at the domain level. So when I click to explore a specific domain, having a HIERARCHY tab that shows that domains hierarchy of groups and the systems they own would be sweet.

miladr commented:
Great job. I recommend following https://c4model.com/ principles for coloring and shaping and also zoom level.
C4Model is already close to Backstage terminology.

Rugvip commented on plugins/explore/src/components/OrganizationExplorerContent/OrganizationDiagram.tsx:
```diff
@@ -0,0 +1,207 @@
+
+// Simplifies the diagram output by hiding the default namespace and kind
+function readableEntityName(
+  ref:
```
These are actually always entities right? Could use `getEntityName(ref).name` instead or at least use that or `ref.name` depending on the shape of `ref`

Rugvip reviewed (commented):
Thanks, I like it! 😁

I think there's some debate here about configuring whether this should be shown or not, but thinking we push that to a more general solution and always display all tabs for now.

A minor nit and I'll let you decide on if you want to tweak the colors etc. according to what @mijailr suggested, but otherwise good to go. Just needs a changeset :grin:

freben commented:
@K-Phoen Just noting that there's a comment above about the need to add a changeset.

K-Phoen commented:
Sorry for the long response time everyone, it's been a busy period for me :/

I addressed @Rugvip's comments: node titles are now generated using `formatEntityRefTitle` (from the `plugin-catalog-react` package`) and I added a changeset.

A lot of good points have been raised in the comments. Calling this diagram an "organization chart" is a bit misleading since like @adamdmharvey noted groups can be sourced from GitHub and will more represent "IAM settings" than a real organization chart.

What's the desired way forward here? Merging as-is and gathering feedback & suggestions for a rework later? Or waiting until we know more?

My preference would be to merge it and see what adopters think – maybe with a few small adjustments (like renaming "Organization" to something more accurate) – but I wouldn't if we chose not to merge straight away :)

andrewthauer commented:
Seems like it's technically the Group entity hierarchy, so maybe just calling it "Group" or "Group Diagram"?

regicsolutions commented:
Domains | Groups | Tools 

Looks right to me...

adamdmharvey commented:
![](https://i.imgflip.com/58cpha.jpg)

freben commented:
deafening agreement!

K-Phoen commented:
:D

Renaming done, and also added a few simple tests.

benjdlambert commented:
Thanks!
